### PR TITLE
2D Polygon Intersection

### PR DESCRIPTION
**Pull request by bmhan12**
This PR:
- Implements an overload `clip()` for 2D polygons using the Sutherland–Hodgman algorithm.
  - host-device decorated to run on device. 
- Generalizes `intersect(plane, segment)` for 2D.
- Moved `StaticArray` from Polygon class into core.

### CODE REVIEW

File: src/axom/primal/geometry/Polygon.hpp [hunk 1]
```diff
@@ -123,6 +91,45 @@ class Polygon
   AXOM_HOST_DEVICE Polygon()
   { }
 
+  /*!
+   * \brief Destructor for Polygon. Suppress CUDA warnings for
+   *        dynamic axom::Array.
+   */
+  AXOM_SUPPRESS_HD_WARN
+  AXOM_HOST_DEVICE
+  ~Polygon() { m_vertices.clear(); }
```
Comment (bmhan12) [inline src/axom/primal/geometry/Polygon.hpp:100]: To get rid of CUDA __host__ __device__ warnings when compiling for the `axom::Array` option, I added `AXOM_SUPPRESS_HD_WARN` here in the Polygon class, as well as to the `clip(Polygon, Polygon)` overload.

I would rather not do this.

Is there a way I can refactor/reorganize the current implementation  to better clear out warnings?

Comment (rhornung67) [reply src/axom/primal/geometry/Polygon.hpp:100]: @bmhan12 are the warnings about calling a device function in host code?

Comment (rhornung67) [reply src/axom/primal/geometry/Polygon.hpp:100]: What happens if you call `clear()` instead of `m_vertices.clear()`?  It looks like you added specialization of the clear method to avoid the warnings.

Comment (bmhan12) [reply src/axom/primal/geometry/Polygon.hpp:100]: Same warnings. The warnings are coming from `clip_impl.hpp`, when the clipPolygonPolygon() function is compiled for device.

**Review by agcapps — APPROVED** [review]
Appreciate the test cases.

File: src/axom/primal/geometry/Polygon.hpp (hunk 1 above)
Comment (publixsubfan) [reply src/axom/primal/geometry/Polygon.hpp:100]: You might be able to move the underlying storage into a templatized base class, that's specialized for `axom::Array` vs `StaticArray`. In the `axom::Array` specialization, you could then define the destructor as host-only, and only have `AXOM_HOST_DEVICE` on the `StaticArray` specialization's destructor.

Comment (bmhan12) [reply src/axom/primal/geometry/Polygon.hpp:100]: @publixsubfan Thanks for the suggestion. I gave it a try, but still needed the `AXOM_SUPPRESS_HD_WARN `macro to suppress warnings.

File: src/axom/core/StaticArray.hpp [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+namespace axom
+{
+/*!
+ * \accelerated
+ * \class StaticArray
+ *
+ * \brief Provides a wrapper for a StackArray with an additional size member.
+ *
+ * \tparam T the type of the values to hold.
```
Comment (rhornung67) [inline src/axom/core/StaticArray.hpp:21]: Is it important to note that the data type 'T' must have a default constructor that can run on a device?

File: src/axom/primal/operators/clip.hpp [hunk 3]
```diff
@@ -95,6 +95,62 @@ Polygon<T, 3> clip(const Triangle<T, 3>& tri, const BoundingBox<T, 3>& bbox)
...
+ *  subject polygon by each plane gives the polygon above that plane.
+ *  Clipping the polygon by a plane involves
+ *  finding new vertices at the intersection of the polygon edges and
+ *  the plane, and removing vertices from the polygon that are below the
+ *  plane.
+ *
+ *
+ * \param [in] subjectPolygon The subject polygon
+ * \param [in] clipPolygon The clip polygon
+ * \param [in] eps The epsilon value
```
Comment (rhornung67) [inline src/axom/primal/operators/clip.hpp:114]:
```suggestion
 * \param [in] eps The epsilon value. Defaults to 1.e-10.
```
You should probably say what the tolerance applies to (could refer to the `clipsPolygonPolygon` methods that uses it if it is documented there).

File: src/axom/primal/operators/detail/clip_impl.hpp [hunk 4]
```diff
@@ -671,6 +671,131 @@ AXOM_HOST_DEVICE Polyhedron<T, NDIMS> clipTetrahedron(
   return poly;
 }
 
+/*!
+ * \brief Clips a subject polygon against a clip polygon and returns the
+ *        resulting polygon
+ *
+ * \param [in] subjectPolygon The subject polygon
+ * \param [in] clipPolygon The clip polygon
+ * \param [in] eps The epsilon value
```
Comment (rhornung67) [inline src/axom/primal/operators/detail/clip_impl.hpp:680]: Same comment as above regarding the `eps` arg.

File: src/axom/primal/operators/detail/intersect_impl.hpp [hunk 5]
```diff
@@ -1100,19 +1100,19 @@ AXOM_HOST_DEVICE bool intersect_plane_bbox(const Plane<T, 3>& p,
 }
 
 /*!
- * \brief Determines if a 3D plane intersects a 3D segment.
- * \param [in] b1 A 3D plane
- * \param [in] b2 A 3D segment
+ * \brief Determines if a plane intersects a segment.
+ * \param [in] b1 A plane
```
Comment (rhornung67) [inline src/axom/primal/operators/detail/intersect_impl.hpp:1104]:
```suggestion
 * \param [in] plane A plane
```

File: src/axom/primal/operators/detail/intersect_impl.hpp [hunk 6]
```diff
@@ -1100,19 +1100,19 @@ AXOM_HOST_DEVICE bool intersect_plane_bbox(const Plane<T, 3>& p,
...
+ * \param [in] b2 A segment
```
Comment (rhornung67) [inline src/axom/primal/operators/detail/intersect_impl.hpp:1105]:
```suggestion
 * \param [in] seg A segment
```

**Review by rhornung67 — COMMENTED** [review]
A few comments to address.

File: src/axom/primal/tests/primal_clip.cpp [hunk 7]
```diff
@@ -1643,6 +1745,184 @@ TEST(primal_clip, tet_plane_intersect_four_edges)
...
+    Polygon2D clipPolygon;
+    clipPolygon.addVertex(Point2D {100, 100});
+    clipPolygon.addVertex(Point2D {300, 100});
+    clipPolygon.addVertex(Point2D {300, 300});
+    clipPolygon.addVertex(Point2D {100, 300});
+
+    Polygon2D poly = axom::primal::clip(subjectPolygon, clipPolygon, EPS);
+    EXPECT_NEAR(poly.signedArea(), 37083.3333333333, EPS);
+    EXPECT_EQ(poly.numVertices(), 10);
+  }
```
Comment (kennyweiss) [inline src/axom/primal/tests/primal_clip.cpp:1923]: 👍 nice test! 

Looks like it's this polygon 
-- https://rosettacode.org/wiki/Sutherland-Hodgman_polygon_clipping#C++
-- https://static.miraheze.org/rosettacodewiki/0/0b/Sutherland-hodgman.png

If so, would it make sense to check that we get these vertices (or something close)?
```
(100, 116.667)
(125, 100)
(275, 100)
(300, 116.667)
(300, 300)
(250, 300)
(200, 250)
(175, 300)
(125, 300)
(100, 250)
```

File: src/axom/primal/operators/detail/clip_impl.hpp [hunk 8]
```diff
@@ -671,6 +671,131 @@ AXOM_HOST_DEVICE Polyhedron<T, NDIMS> clipTetrahedron(
...
+
+      if(cur_p_orientation == ON_POSITIVE_SIDE)
+      {
+        // Handles the edge case of 3 consecutive vertices with orientations
+        // ON_POSITIVE_SIDE, ON_BOUNDARY, ON_POSITIVE. Default algorithm
+        // check (prev_p_orientation != ON_POSITIVE_SIDE) results in the
+        // vertex on the boundary being added twice.
+        if(prev_p_orientation == ON_NEGATIVE_SIDE ||
+           (prev_p_orientation == ON_BOUNDARY &&
+            intersecting_point != outputList[outputList.numVertices() - 1]))
```
Comment (kennyweiss) [inline src/axom/primal/operators/detail/clip_impl.hpp:783]: Are there ever cases where this will be called and the curves did not intersect (e.g. intersecting point was default initialized)?

It seems like we might want to move the intersect call into this check as follows.
(I think the following change should be equivalent and more efficient, but please check):
```cpp

      const int cur_p_orientation = plane.getOrientation(current_point, eps);
      const int prev_p_orientation = plane.getOrientation(prev_point, eps);

      if(cur_p_orientation == ON_POSITIVE_SIDE)
      {
        // Handles the edge case of 3 consecutive vertices with orientations
        // ON_POSITIVE_SIDE, ON_BOUNDARY, ON_POSITIVE. Default algorithm
        // check (prev_p_orientation != ON_POSITIVE_SIDE) results in the
        // vertex on the boundary being added twice.
        if(prev_p_orientation == ON_NEGATIVE_SIDE ||
           (prev_p_orientation == ON_BOUNDARY)
        { 
          T seg_param;
          SegmentType subject_edge(prev_point, current_point);

           if(intersect(plane, subject_edge, seg_param))
           {
              const auto intersecting_point = subject_edge.at(seg_param);
              if(intersecting_point != outputList[outputList.numVertices() - 1]))
              {
                 outputList.addVertex(intersecting_point);
              }
           }
         }
         outputList.addVertex(current_point);
      }
      else if(prev_p_orientation == ON_POSITIVE_SIDE)
      {
        outputList.addVertex(intersecting_point);
      }
```

Comment (kennyweiss) [reply src/axom/primal/operators/detail/clip_impl.hpp:783]: Also, do we have tests that cover all these cases?

**Review by kennyweiss — APPROVED** [review]
Thanks @bmhan12!

Please don't forget to update the Release-Notes

**Review by rhornung67 — APPROVED** [review]

File: src/axom/primal/operators/clip.hpp [hunk 9]
```diff
@@ -95,6 +95,63 @@ Polygon<T, 3> clip(const Triangle<T, 3>& tri, const BoundingBox<T, 3>& bbox)
...
+ *             Defaults to false.
+ *
+ * \return A polygon of the subject polygon clipped against the clip polygon.
+ *
+ * \note Function is based off the Sutherland–Hodgman algorithm.
+ *
+ * \warning Polygons with static array types must have enough vertices
+ *          preallocated for the output polygon. It is recommended that
+ *          MAX_VERTS >= subjectPolygon.numVertices() + clipPolygon.numVertices()
+ *          for the output polygon with the largest possible vertex count.
```
[the remark this answers is not in the document]
Comment (bmhan12) [reply src/axom/primal/operators/clip.hpp]: > Should we indicate what will happen if there are not enough vertices, or how the user would know that this was the case?

Updated doxygen warning for expected behavior if not enough vertices are provided (output polygon will be missing vertices).

> Also, should we indicate (or enforce) that when running on device, the ARRAY_TYPE is static?

Added a `SLIC_ASSERT` that makes the recommended max size mandatory for static polygons.

File: src/axom/primal/operators/detail/clip_impl.hpp (hunk 8 above)
Comment (bmhan12) [reply src/axom/primal/operators/detail/clip_impl.hpp:751]: > Are there ever cases where this will be called and the curves did not intersect (e.g. intersecting point was default initialized)?

No, there will always be an intersection when the if condition is satisfied at this point.
The main issue was the edge case I mentioned in the comments, which can result in a point being added twice in that specific scenario.

> It seems like we might want to move the intersect call into this check as follows.
(I think the following change should be equivalent and more efficient, but please check):

I gave it a try, but there is a compilation failure as `intersection_point` is not defined at the `else-if` block:
```
      else if(prev_p_orientation == ON_POSITIVE_SIDE)
      {
        outputList.addVertex(intersecting_point); // intersecting_point not defined
      }
```

I left the branching structure as-is, since the intersection point needs to be known by both branch paths.
The Sutherland–Hodgman algorithm should guarantee there is an intersection point in the `else-if` case, as when teh condition is satisfied the current point is either on the negative side or on the edge, while the previous point is on the positive side.

> Also, do we have tests that cover all these cases?

Yes, all branch paths are covered, including a test for the specific edge case mentioned.
